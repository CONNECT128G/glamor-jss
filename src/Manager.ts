import { Rule, Style, SheetsRegistry, StyleSheet } from 'jss'
import { jss } from './index'

const IS_DEV = process.env.NODE_ENV !== 'production'

export const MAX_RULES = 65534

interface Options {
  sheetPrefix: string
  classNamePrefix: string
}

const defaultOptions = {
  sheetPrefix: 'glamor-jss',
  classNamePrefix: 'css',
} as Options

export default class Manager {
  registry = new SheetsRegistry()
  currentSheet: StyleSheet | null = null
  rulesCount = 0
  sheetCount = 0
  options: Options

  constructor(options?: Options) {
    this.options = {
      ...defaultOptions,
      ...options,
    }
  }

  reset = () => {
    this.registry.reset()
    this.currentSheet = null
  }

  createSheet = (): StyleSheet => {
    const { sheetPrefix } = this.options

    const sheet = jss.createStyleSheet<string>(
      {},
      {
        generateId: (rule: Rule) =>
          `${this.options.classNamePrefix}-${rule.key}`,
        meta: `${sheetPrefix}-${this.sheetCount++}`,
      }
    )
    this.registry.add(sheet)

    return sheet
  }

  getSheet = () => {
    if (!this.currentSheet) {
      this.currentSheet = this.createSheet()
    }
    return this.currentSheet
  }

  addRule = (hash: string | number, declarations: Style, options?: any) => {
    const sheet = this.getSheet()

    // Detatch and attach again to make Chrome Dev Tools working
    // Similar to `speedy` from glamor: https://github.com/threepointone/glamor#speedy-mode
    if (IS_DEV) sheet.detach()
    const rule = sheet.addRule(String(hash), declarations, options)
    // https://blogs.msdn.microsoft.com/ieinternals/2011/05/14/stylesheet-limits-in-internet-explorer/
    if (++this.rulesCount % MAX_RULES === 0) {
      this.currentSheet = this.createSheet()
    }
    if (IS_DEV) sheet.attach()

    return rule
  }
}
