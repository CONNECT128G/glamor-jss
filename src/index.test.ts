import css, { renderToString, reset } from './index'
import Manager, { MAX_RULES } from './Manager'

describe('css', () => {
  beforeEach(reset)

  test('Simple styles', () => {
    const style = css({ width: 100, height: 100 })
    expect(renderToString()).toMatchInlineSnapshot(`
      ".css-5638800070770, [data-css-5638800070770] {
        width: 100px;
        height: 100px;
      }"
    `)
    expect(style.toString()).not.toBeUndefined()

    expect(Object.keys({ ...style })[0]).toEqual(
      expect.stringMatching(/^data-css/)
    )

    expect(css()).toBeUndefined()
  })

  test('Keyframes work without collision', () => {
    css.keyframes('fade', { from: { opacity: 0 }, to: { opacity: 1 } })
    css.keyframes('fade', { from: { opacity: 0 }, to: { opacity: 1 } })
    css.keyframes({ from: { opacity: 0 }, to: { opacity: 1 } })
    css.keyframes({ from: { opacity: 0 }, to: { opacity: 1 } })
    expect(renderToString()).toMatchSnapshot()
  })

  test('Falsy values', () => {
    css({ width: null, height: undefined, minWidth: false, ':hover': {} }, null)
    css([null, {}, []])
    css({ ':after': { width: false && 100, height: 100 } })

    expect(renderToString()).toMatchInlineSnapshot(`""`)
  })

  test('Complex styles', () => {
    const bold = css({
      fontWeight: 'bold',
    })
    css(
      [
        { color: 'lightgoldenrodyellow' },
        { height: 150, '&:hover': { color: 'khaki', background: 'yellow' } },
      ],
      {
        border: '1px solid mediumaquamarine',
        width: 350,
        borderRadius: '50%',
      },
      [
        {
          color: 'papayawhip',
          '@media (min-width: 800px)': { background: 'peachpuff' },
        },
        {
          '&:hover': {
            color: 'orchid',
          },
        },
      ],
      bold
    )

    expect(renderToString()).toMatchSnapshot()
  })

  test('Nested functional styles', () => {
    const activeStyle = {
      color: 'peachpuff',
      '&:before': {
        top: 10,
      },
    }

    const style = (hover: boolean) =>
      css({
        position: 'relative',
        color: 'gray',
        '&:before': {
          position: 'absolute',
          content: `""`,
          top: 0,
        },
        '&:hover': hover && activeStyle,
      })

    const styles = (toggle: boolean) =>
      css(style(toggle), toggle && activeStyle)

    expect(styles(true)).toMatchInlineSnapshot(`
      Object {
        "data-css-3275816868020": "",
      }
    `)
    expect(styles(false)).toMatchInlineSnapshot(`
      Object {
        "data-css-8455158449034": "",
      }
    `)
    expect(renderToString()).toMatchSnapshot()
  })

  test('Overwrite styles', () => {
    css(
      css({ color: 'mediumaquamarine' }),
      { color: 'peachpuff' },
      css({ color: 'lightcoral' }),
      { color: 'papayawhip' },
      css({ color: 'thistle' })
    )

    expect(renderToString()).toMatchSnapshot()
  })

  test('Multiple styles', () => {
    const props = {
      ...css({ color: 'peachpuff' }),
      ...css({ width: '100vw' }),
    }

    expect(Object.keys(props)).toMatchSnapshot()
  })

  test('Cache styles', () => {
    const complex = [
      [
        { color: 'lightgoldenrodyellow' },
        { height: 150, '&:hover': { color: 'khaki' } },
      ],
      {
        border: '1px solid mediumaquamarine',
        width: 350,
        borderRadius: '50%',
      },
      [
        {
          color: 'papayawhip',
          '@media (min-width: 800px)': { background: 'peachpuff' },
        },
      ],
    ]

    expect(css({ color: 'red' })).toBe(css({ color: 'red' }))
    expect(css({ width: (y: number) => y + 1 })).toEqual(
      css({ width: (y: number) => y + 1 })
    )
    expect(css(complex)).toBe(css(complex))
  })

  test('Max rules', () => {
    const man = new Manager()
    man.rulesCount = MAX_RULES - 1
    man.addRule('test', { color: 'red' })

    expect((man.registry as any).registry).toHaveLength(2)
  })
})
