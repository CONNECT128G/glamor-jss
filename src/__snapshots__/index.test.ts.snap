// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css Complex styles 1`] = `
".css-12483742431659, [data-css-12483742431659] {
  font-weight: bold;
}
.css-16140790311946, [data-css-16140790311946] {
  color: papayawhip;
  width: 350px;
  border: 1px solid mediumaquamarine;
  height: 150px;
  font-weight: bold;
  border-radius: 50%;
}
.css-16140790311946:hover, [data-css-16140790311946]:hover {
  color: orchid;
  background: yellow;
}
@media (min-width: 800px) {
  .css-16140790311946 {
    background: peachpuff;
  }
}"
`;

exports[`css Keyframes work without collision 1`] = `
"@keyframes css-keyframes-fade-0 {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes css-keyframes-fade-1 {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes css-keyframes-animation-2 {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes css-keyframes-animation-3 {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}"
`;

exports[`css Multiple styles 1`] = `
Array [
  "data-css-7696658414362",
  "data-css-15533376699467",
]
`;

exports[`css Nested functional styles 3`] = `
".css-2084834786935, [data-css-2084834786935] {
  color: gray;
  position: relative;
}
.css-2084834786935:before, [data-css-2084834786935]:before {
  top: 0;
  content: \\"\\";
  position: absolute;
}
.css-2084834786935:hover, [data-css-2084834786935]:hover {
  color: peachpuff;
}
.css-2084834786935:hover:before, [data-css-2084834786935]:hover:before {
  top: 10px;
}
.css-3275816868020, [data-css-3275816868020] {
  color: peachpuff;
  position: relative;
}
.css-3275816868020:before, [data-css-3275816868020]:before {
  top: 10px;
  content: \\"\\";
  position: absolute;
}
.css-3275816868020:hover, [data-css-3275816868020]:hover {
  color: peachpuff;
}
.css-3275816868020:hover:before, [data-css-3275816868020]:hover:before {
  top: 10px;
}
.css-10007073420595, [data-css-10007073420595] {
  color: gray;
  position: relative;
}
.css-10007073420595:before, [data-css-10007073420595]:before {
  top: 0;
  content: \\"\\";
  position: absolute;
}
.css-8455158449034, [data-css-8455158449034] {
  color: gray;
  position: relative;
}
.css-8455158449034:before, [data-css-8455158449034]:before {
  top: 0;
  content: \\"\\";
  position: absolute;
}"
`;

exports[`css Overwrite styles 1`] = `
".css-1113385277541, [data-css-1113385277541] {
  color: mediumaquamarine;
}
.css-7242087029549, [data-css-7242087029549] {
  color: lightcoral;
}
.css-14164733049051, [data-css-14164733049051] {
  color: thistle;
}
.css-13164275224355, [data-css-13164275224355] {
  color: thistle;
}"
`;
